packages/web-components/fast-components/src/color-2/palette.ts

```diff
@@ -0,0 +1,126 @@
+import {
+    clamp,
+    ComponentStateColorPalette,
+    parseColorHexRGB,
+} from "@microsoft/fast-colors";
+import { Swatch, SwatchRGB } from "./swatch";
+import { directionByMode } from "./utilities/direction-by-mode";
+import { contrast } from "./utilities/relative-luminance";
+
+export interface Palette<T extends Swatch = Swatch> {
+    readonly source: T;
+    readonly swatches: ReadonlyArray<T>;
+
+    /**
+     * Returns a swatch from the palette that most closely matches
+     * the contrast ratio provided to a provided reference.
+     *
+     * @remarks
+     * We need to know where to start searching
+     * and which direction to search in - isDarkMode?
+     */
+    colorContrast(reference: Swatch, contrast: number): Swatch;
+
+    /**
+     * Returns the index of the palette that most closely matches
+     * the relativeLuminance of the provided swatch
+     */
+    closestIndexOf(reference: Swatch): number;
+
+    /**
+     * Gets a swatch by index. Index is clamped to the limits
+     * of the palette so a Swatch will always be returned.
+     */
+    get(index: number): Swatch;
+}
+
+export class PaletteRGB implements Palette<SwatchRGB> {
+    public readonly source: SwatchRGB;
+    public readonly swatches: ReadonlyArray<SwatchRGB>;
+    private lastIndex: number;
+    private reversedSwatches: ReadonlyArray<SwatchRGB>;
+    constructor(source: SwatchRGB) {
+        this.swatches = Object.freeze(
+            new ComponentStateColorPalette({
+                baseColor: source,
+            }).palette
+                .map(x => parseColorHexRGB(x.toStringHexRGB())!) // TODO: we need lower fidelity
+                .map(x => new SwatchRGB(x.r, x.g, x.b))
+        );
+
+        this.reversedSwatches = Object.freeze([...this.swatches].reverse());
+        this.lastIndex = this.swatches.length - 1;
+    }
+
+    public colorContrast(reference: Swatch, contrastTarget: number): SwatchRGB {
+        const referenceIndex = this.closestIndexOf(reference);
+        let startIndex: number;
+        let source: ReadonlyArray<SwatchRGB>;
+        const direction = directionByMode(reference);
```

Comment:
This represents a change from the current functionality, but I think we should leave the decision about the direction to the individual use case / recipe. The notion of light or dark mode is for the UI as a whole, and it's harder to deal with on an individual color. We at least need a `preferred direction` or some way to look for consistency across an experience. A real example where this comes up is wanting to keep white text over various accent colors. If the background supports it, even if black would have _more_ contrast, if white _passes_ it should be used.

Thinking of this more from an lower-level class perspective, if I'm asking for a lower contrast value like 2:1 I'm probably using it in some way where I am less concerned with light / dark mode and would want to specify the direction.

Reply:
I might be missing context, but I do have an idea of what @bheston is getting at with this:

> A real example where this comes up is wanting to keep white text over various accent colors. If the background supports it, even if black would have more contrast, if white passes it should be used

In the above scenario, is it too much to ask them to provide an alternative recipe? Do we need a binary switch here? If you want to change how a color is resolved, I think the system now provides a method for that with DI. But I'm not sure we need to expose that. That type of configuration as a part of the core color library becomes expensive and something that needs to be resolved every time. Is there a reason that wholesale replacing the recipe in the case above is not reasonable? Full disclosure, I think that's currently the expectation from the team asking me for the above behavior. I don't know that it's universal enough to inform the recipes themselves or to take the potential perf hit of checking every time.

Reply:
We've had feedback in Fluent UI to update the recipe for our accent buttons to defaulting to *white* rather than black when the above is true, but I don't think it's necessary in each recipe personally.